javaslang/src/main/java/javaslang/Lazy.java

```diff
@@ -135,14 +133,11 @@ private Lazy(Supplier<? extends T> supplier) {
      */
     @Override
     public T get() {
-        // using a local var speeds up the double-check idiom by 25%, see Effective Java, Item 71
-        Supplier<? extends T> tmp = supplier;
-        if (tmp != null) {
+        if ((value == null) && (supplier != null)) {
```

Comment:
@shipilev In our case that might be ok. We first read `value` because it is a fast read. We can have three cases then.

1. value was not written by any thread yet
2. read reflects the written value, i.e. it is visible to the thread.
3. written value is not visible to the current thread

In case 1. `value` is null and will be computed. After that it is visible to the current thread and will be returned by `get()`.

In the second case the write to `value` is visible. Then we are finished and can safely return it.

In the third case the write to `value` is not visible, it seems to be still null. Then `supplier` is read and it turns out that we already computed the value. Because of the volatile read of `supplier` our `value` is now ✨automagically✨ visible to the current thread and it can be safely returned by `get()`.

In other words we use `value == null` as a shortcut and we can detect not-visible writes by having the volatile 'afterburner' read of `supplier`.

Benchmarks say that is faster in most cases.